File: domain/src/main/java/io/prizy/domain/contest/model/Pack.java
package io.prizy.domain.contest.model;

import java.time.Instant;
import java.util.UUID;

import lombok.Builder;
import lombok.With;


public sealed interface Pack {

  UUID id();

  String name();

  Integer lastWinnerPosition();

  Integer firstWinnerPosition();

  @With
  @Builder
  record Product(
    UUID id,
    String name,
    Integer lastWinnerPosition,
    Integer firstWinnerPosition,
    Integer quantity,
    Double value,
    String assetId
  ) implements Pack {
  }

  @With
  @Builder
  record Coupon(
    UUID id,
    String name,
    Integer lastWinnerPosition,
    Integer firstWinnerPosition,
    String code,
    Instant expiration
  ) implements Pack {
  }
}

File: domain/src/main/kotlin/io/prizy/domain/contest/model/Pack.kt
package io.prizy.domain.contest.model

import java.time.Instant
import java.util.UUID

sealed interface Pack {

  val id: UUID
  val name: String
  val lastWinnerPosition: Int
  val firstWinnerPosition: Int

  data class Product(
    override val id: UUID,
    override val name: String,
    override val lastWinnerPosition: Int,
    override val firstWinnerPosition: Int,
    val quantity: Int,
    val value: Float,
    val mediaId: String
  ) : Pack

  data class Coupon(
    override val id: UUID,
    override val name: String,
    override val lastWinnerPosition: Int,
    override val firstWinnerPosition: Int,
    val code: String,
    val expiration: Instant,
  ) : Pack
}



